File: server/recbridge/recorder/epg.py
"""Decoder (and test encoder) for the EPG files a Sony BDZ recorder serves on port 60151.

Format (as served by a BDZ-FBT4100; see docs/epg-format.md):
  raw file  = XOR 0x9D over a concatenation of zlib streams, one stream per service.
  stream    = "@SRV" record: 156-byte header, then "@DAY" blocks holding "@EVT" blocks.
  times     = seconds since 1970-01-01 00:00 *JST* (unix time + 32400).
"""
from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone

JST = timezone(timedelta(hours=9))
_JST_OFFSET = 32400
_XOR = 0x9D
_HEADER_LEN = 156
_EVT_HEADER_LEN = 56


@dataclass
class Program:
    service_id: int
    event_id: int
    start: datetime
    end: datetime
    title: str = ""
    description: str = ""
    extended: str = ""
    genres: list[tuple[int, int]] = field(default_factory=list)  # (level1, level2) nibbles
    copy_control: int = 0
    parental_rating: int = 0
    ref_service_id: int | None = None  # set for simulcast references on sub-channels
    ref_event_id: int | None = None

    @property
    def is_reference(self) -> bool:
        return self.ref_event_id is not None

    @property
    def duration_sec(self) -> int:
        return int((self.end - self.start).total_seconds())


@dataclass
class Service:
    service_id: int
    name: str
    programs: list[Program] = field(default_factory=list)


def _be16(b: bytes, i: int) -> int:
    return (b[i] << 8) | b[i + 1]


def _be32(b: bytes, i: int) -> int:
    return struct.unpack_from(">I", b, i)[0]


def _ts(v: int) -> datetime:
    return datetime.fromtimestamp(v - _JST_OFFSET, JST)


# ARIB additional symbols arrive as private-use code points (mapping: docs/epg-format.md).
ARIB_SYMBOLS = {
    "\ue0fd": "[手]", "\ue0fe": "[字]", "\ue180": "[デ]", "\ue182": "[二]", "\ue183": "[多]", "\ue184": "[解]",
    "\ue185": "[SS]", "\ue18c": "[映]", "\ue192": "[再]", "\ue193": "[新]", "\ue195": "[終]", "\ue196": "[生]",
}


def _clean(b: bytes) -> str:
    """UTF-8 text from the recorder: NUL padding and C0 controls dropped, ARIB symbols spelled out, unknown private-use chars removed."""
    out = []
    for ch in b.decode("utf-8", "replace"):
        if ch in ARIB_SYMBOLS:
            out.append(ARIB_SYMBOLS[ch])
        elif "\ue000" <= ch <= "\uf8ff" or (ch < " " and ch != "\n"):
            continue
        else:
            out.append(ch)
    return "".join(out).strip()


def _unts(dt: datetime) -> int:
    return int(dt.timestamp()) + _JST_OFFSET


def split_streams(raw: bytes) -> list[bytes]:
    """XOR-decode and inflate every concatenated zlib stream."""
    x = bytes(b ^ _XOR for b in raw)
    out: list[bytes] = []
    pos = 0
    while pos < len(x):
        d = zlib.decompressobj()
        data = d.decompress(x[pos:])
        used = len(x) - pos - len(d.unused_data)
        if used <= 0:
            break
        out.append(data)
        pos += used
    return out


def parse_service(rec: bytes) -> Service:
    if rec[:4] != b"@SRV":
        raise ValueError("not an @SRV record")
    service_id = _be16(rec, 8)
    name_len = _be16(rec, 26)
    name = _clean(rec[28:28 + name_len])
    svc = Service(service_id=service_id, name=name)
    p = _HEADER_LEN
    while p + 16 <= len(rec) and rec[p:p + 4] == b"@DAY":
        block_len = _be32(rec, p + 8)
        e = p + 16
        end = p + block_len
        while e + 12 <= end and rec[e:e + 4] == b"@EVT":
            event_len = _be16(rec, e + 8)
            if event_len < 28:
                break
            flags = rec[e + 10]
            pr = Program(service_id=service_id, event_id=_be16(rec, e + 6),
                         start=_ts(_be32(rec, e + 16)), end=_ts(_be32(rec, e + 20)))
            if (flags >> 6) & 1 and not (flags >> 5) & 1:
                # reference form (simulcast on a sub-channel): only start/end plus the parent service/event
                pr.ref_service_id = _be16(rec, e + 24)
                pr.ref_event_id = _be16(rec, e + 26)
            else:
                # three 2-byte slots: content nibbles, then user nibbles (0xFF when the slot is in use)
                pr.genres = [(rec[e + 30 + 2 * i] >> 4, rec[e + 30 + 2 * i] & 0xF) for i in range(3)
                             if rec[e + 31 + 2 * i] != 0 or rec[e + 30 + 2 * i] != 0]
                pr.copy_control = (rec[e + 40] & 0x0C) >> 2
                rating = rec[e + 41] & 0x1F
                pr.parental_rating = 0 if rating < 4 else rating - 3
                n_title, n_desc, title_field, desc_end, n_ext = (_be16(rec, e + 44 + 2 * i) for i in range(5))
                t = e + _EVT_HEADER_LEN
                pr.title = _clean(rec[t:t + n_title])
                pr.description = _clean(rec[t + title_field:t + title_field + n_desc])
                pr.extended = _clean(rec[t + desc_end:t + desc_end + n_ext])
            svc.programs.append(pr)
            e += event_len
        p += block_len
    return svc


def decode_epg_file(raw: bytes) -> list[Service]:
    return [parse_service(s) for s in split_streams(raw)]


# --- encoder: used by tests to build fixtures and to prove the layout is understood both ways ---

def _encode_event(pr: Program) -> bytes:
    if pr.is_reference:
        body = bytearray(b"@EVT" + b"\x04\x00")
        body += struct.pack(">HHB", pr.event_id, 0, 0x44) + b"\x00" * 5
        body += struct.pack(">II", _unts(pr.start), _unts(pr.end))
        body += struct.pack(">HH", pr.ref_service_id or 0, pr.ref_event_id or 0)
        body += b"\x00" * 4
        struct.pack_into(">H", body, 8, len(body))
        return bytes(body)
    title, desc, ext = pr.title.encode(), pr.description.encode(), pr.extended.encode()
    title_field = len(title) + 2
    desc_end = title_field + len(desc) + 2
    body = bytearray(b"@EVT" + b"\x04\x00")
    body += struct.pack(">HHB", pr.event_id, 0, 0x64) + b"\x08\x00\x00\x00\x01"
    body += struct.pack(">II", _unts(pr.start), _unts(pr.end)) + b"\x00" * 6
    genres = list(pr.genres)[:3]
    for l1, l2 in genres:
        body += bytes([(l1 << 4) | l2, 0xFF])
    body += b"\x00\x00" * (3 - len(genres))
    body += b"\x00" * 4 + bytes([pr.copy_control << 2, 0 if pr.parental_rating == 0 else pr.parental_rating + 3]) + b"\x00" * 2
    body += struct.pack(">HHHHH", len(title), len(desc), title_field, desc_end, len(ext)) + b"\x00" * 2
    assert len(body) == _EVT_HEADER_LEN
    body += title + b"\x00\x00" + desc + b"\x00\x00" + ext + b"\x00\x00"
    struct.pack_into(">H", body, 8, len(body))
    return bytes(body)


def encode_service(svc: Service) -> bytes:
    """Build one inflated @SRV record (no XOR/zlib)."""
    name = svc.name.encode()
    hdr = bytearray(b"@SRV" + struct.pack(">HHH", 3, 0, svc.service_id) + b"\x00\x08" + b"\x00" * 4)
    hdr += b"\x00" * (26 - len(hdr)) + struct.pack(">H", len(name)) + name
    hdr += b"\x00" * (_HEADER_LEN - len(hdr))
    days: dict[datetime, list[Program]] = {}
    for pr in svc.programs:
        day = pr.start.astimezone(JST).replace(hour=0, minute=0, second=0, microsecond=0)
        days.setdefault(day, []).append(pr)
    body = bytearray()
    for day, prs in sorted(days.items()):
        events = b"".join(_encode_event(p) for p in prs)
        body += b"@DAY" + struct.pack(">IIB", _unts(day), 16 + len(events), len(prs)) + b"\x00" * 3 + events
    struct.pack_into(">I", hdr, 12, len(hdr) + len(body) - 4)
    return bytes(hdr) + bytes(body)


def encode_epg_file(services: list[Service]) -> bytes:
    raw = b"".join(zlib.compress(encode_service(s)) for s in services)
    return bytes(b ^ _XOR for b in raw)

File: app/RecorderKit/Sources/RecorderKit/Epg.swift
import Foundation

/// Decoder for the `EPG_*_FILE.dat` files the recorder serves on its media server port.
///
/// The file is XOR 0x9D over a run of zlib streams, one per service. Each stream is an `@SRV` record: a
/// 156-byte header, then `@DAY` blocks holding `@EVT` blocks. Times are seconds since 1970-01-01 00:00 *JST*.
/// See docs/epg-format.md; docs/port/epg-sample.json pins the result down.
public enum Epg {
    static let xorKey: UInt8 = 0x9D
    static let jstEpochOffset = 32400
    static let serviceHeaderLength = 156
    static let eventHeaderLength = 56

    public static func decode(_ data: Data) throws -> [GuideService] {
        try splitStreams(data).compactMap { try? parseService($0) }
    }

    /// Un-XORs the file and inflates every stream in it.
    static func splitStreams(_ data: Data) throws -> [Data] {
        var bytes = [UInt8](data)
        for index in bytes.indices { bytes[index] ^= xorKey }

        var streams: [Data] = []
        var position = 0
        while position < bytes.count {
            let (output, consumed) = try Inflate.first(Array(bytes[position...]))
            if consumed <= 0 { break }
            streams.append(output)
            position += consumed
        }
        return streams
    }

    static func parseService(_ record: Data) throws -> GuideService {
        let bytes = Bytes([UInt8](record))
        guard bytes.marker("@SRV", at: 0) else { throw GuideError.notAServiceRecord }

        let serviceID = bytes.be16(8)
        let nameLength = bytes.be16(26)
        var programs: [GuideProgram] = []

        var day = serviceHeaderLength
        while day + 16 <= bytes.count, bytes.marker("@DAY", at: day) {
            let blockLength = bytes.be32(day + 8)
            guard blockLength > 16 else { break }
            let blockEnd = day + blockLength

            var event = day + 16
            while event + 12 <= blockEnd, bytes.marker("@EVT", at: event) {
                let eventLength = bytes.be16(event + 8)
                guard eventLength >= 28 else { break }
                programs.append(parseEvent(bytes, at: event, serviceID: serviceID))
                event += eventLength
            }
            day += blockLength
        }
        return GuideService(serviceID: serviceID, name: bytes.text(28, nameLength), programs: programs)
    }

    private static func parseEvent(_ bytes: Bytes, at event: Int, serviceID: Int) -> GuideProgram {
        var program = GuideProgram(
            serviceID: serviceID,
            eventID: bytes.be16(event + 6),
            start: date(bytes.be32(event + 16)),
            end: date(bytes.be32(event + 20)),
            title: "", summary: "", extended: "",
            genres: [], copyControl: 0, parentalRating: 0
        )

        let flags = bytes.byte(event + 10)
        let isReference = (flags >> 6) & 1 == 1 && (flags >> 5) & 1 == 0
        if isReference {
            // A simulcast on a sub-channel: start and end, plus which programme on the parent service it is.
            program.referenceServiceID = bytes.be16(event + 24)
            program.referenceEventID = bytes.be16(event + 26)
            return program
        }

        // Three two-byte slots: the content nibbles, then the user nibbles.
        program.genres = (0..<3).compactMap { slot in
            let content = bytes.byte(event + 30 + 2 * slot)
            let user = bytes.byte(event + 31 + 2 * slot)
            guard content != 0 || user != 0 else { return nil }
            return Genre(level1: content >> 4, level2: content & 0xF)
        }
        program.copyControl = (bytes.byte(event + 40) & 0x0C) >> 2
        let rating = bytes.byte(event + 41) & 0x1F
        program.parentalRating = rating < 4 ? 0 : rating - 3

        let titleLength = bytes.be16(event + 44)
        let summaryLength = bytes.be16(event + 46)
        let summaryOffset = bytes.be16(event + 48)
        let extendedOffset = bytes.be16(event + 50)
        let extendedLength = bytes.be16(event + 52)
        let text = event + eventHeaderLength
        program.title = bytes.text(text, titleLength)
        program.summary = bytes.text(text + summaryOffset, summaryLength)
        program.extended = bytes.text(text + extendedOffset, extendedLength)
        return program
    }

    static func date(_ recorderSeconds: Int) -> Date {
        Date(timeIntervalSince1970: TimeInterval(recorderSeconds - jstEpochOffset))
    }
}

/// Reads big-endian fields without trapping on a short or malformed record.
struct Bytes {
    private let bytes: [UInt8]

    init(_ bytes: [UInt8]) { self.bytes = bytes }

    var count: Int { bytes.count }

    func byte(_ index: Int) -> Int {
        bytes.indices.contains(index) ? Int(bytes[index]) : 0
    }

    func be16(_ index: Int) -> Int {
        byte(index) << 8 | byte(index + 1)
    }

    func be32(_ index: Int) -> Int {
        byte(index) << 24 | byte(index + 1) << 16 | byte(index + 2) << 8 | byte(index + 3)
    }

    func marker(_ marker: String, at index: Int) -> Bool {
        let expected = [UInt8](marker.utf8)
        guard index >= 0, index + expected.count <= bytes.count else { return false }
        return Array(bytes[index..<index + expected.count]) == expected
    }

    /// UTF-8 text of a field, with the recorder's padding and ARIB symbols dealt with.
    func text(_ index: Int, _ length: Int) -> String {
        guard length > 0, index >= 0, index < bytes.count else { return "" }
        let end = min(index + length, bytes.count)
        return Arib.clean(String(decoding: bytes[index..<end], as: UTF8.self))
    }
}
